src/main/ruby/core/dir_glob.rb

```diff
@@ -280,6 +280,10 @@ def self.path_split(str)
     end
 
     def self.single_compile(glob, flags=0)
+      if NO_GLOB_META_CHARS.match?(glob)
+        return ConstantEntry.new nil, flags, glob
+      end
```

Comment:
Hmm... I think you're right. I'll add a spec and update.

Reply:
The `RootDirectory` case is fine. All it does is set the start of the filename to '/' and delegates to the next entry in the chain. `DirectoriesOnly` is trickier. Rather than do a simple file existence check, it attempts to verify the file not only exists but is a directory. I can retain the same behavior (i.e., checking if the last byte is '/'), but I've been unable to induce a failure case.

The best I could think of was creating a regular file on Windows that ends with '/', but where the directory separator is '\'. However, standard Windows tools wouldn't allow me to create a normal file that ends with '/'. Are you aware of any file system or OS where this is actually a problem?

Reply:
I think DirectoriesOnly applies if the last character of the pattern is `/`.
For instance:
```ruby
[6] pry(main)> Dir["notes.txt/"]
=> [] # notes.txt is a file
[7] pry(main)> Dir["prefix/"]
=> ["prefix/"]
```

Reply:
It also seems to filter directories with, e.g.: `Dir["*/"]`.

Reply:
Interestingly it seems `File.exist?` also filter directories if ending with a final `/`. I wonder if that's OS-specific or portable. Probably good enough if it works on Linux+macOS, but would be good to add a spec. Another way is to not shortcut if the last character is `/`.

Reply:
I added a spec in 5c9e012e429965ef243082337c21ebea2c8be1a2, but it didn't really demonstrate much of anything.

Reply:
The `Dir["*/"]` case would be bypassed by my short-circuit because of the presence of the `*` character.

Reply:
Another option is to push the `getbyte(-1) == 47` check into `ConstantEntry`, so its `call` method looks like:

```
def call(matches, parent)
  path = path_join(parent, @name)

  check = getbyte(-1) == 47 ? Dir.exist?(path) : File.exist?(path)
  if check
    matches << path
  end
end
```

Thoughts?

This changes the directory check that `DirectoriesOnly` performs, but I think `Dir.exist?` is clearer than `File.exist?("#{path}/.")`. `Dir.exist?` is currently more expensive, but we can fix that.

Reply:
It sounds fine too but I think a check next to `NO_GLOB_META_CHARS.match?(glob)` is clearer, so the ConstantEntry stays as simple as possible. I would guess patterns ending in `/` are not very common.

But I wonder why `Dir.glob` is used instead of File.exist?/Dir.exist? though, do you know?

Reply:
RubyGems does globbing using `{}` in various path components to choose between multiple candidates. The `compile` method resolves each of those to a name that doesn't contain metachars and falls over to the `single_compile` method. The old logic processed such paths fine -- it's just unnecessarily expensive (and measurably so).